File: backend/app/schemas/xai.py
"""XAI 可解释体系 — 推理路径可视化 Schema

诚实地标注为「结果分解」而非「推理过程」：
- Qwen-VL 是一次性调用返回所有信息，不存在真实的逐步推理
- 本模块将 AI 输出结果按维度拆解为树形结构，每个节点标注来源类型
"""

from __future__ import annotations

from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


# ── 来源类型 ──────────────────────────────────────────────

class SourceType(str, Enum):
    """证据来源类型（四色编码）"""
    COMPUTED = "computed"           # 绿色 — 可验证的计算结果（阈值判定、difflib 匹配度）
    MODEL_OUTPUT = "model-output"   # 蓝色 — AI 模型推断输出（Qwen-VL 标签、LLM 文本）
    RETRIEVED = "retrieved"         # 橙色 — 知识库检索结果（RAG 片段、语义相似度匹配）
    RULE_BASED = "rule-based"       # 紫色 — 规则判定（印章条件、段位计算）


# ── 推理追踪树 ────────────────────────────────────────────

class TraceNode(BaseModel):
    """推理追踪树节点 — ECharts tree 系列的数据格式

    每个节点代表 AI 输出结果的一个维度分解。
    """
    name: str = Field(..., description="节点标签，如「品类判定」「损伤分析」")
    source_type: SourceType = Field(..., description="证据来源类型")
    confidence: Optional[float] = Field(default=None, ge=0, le=1, description="置信度 0-1，仅叶节点或有明确分数的节点")
    detail: Optional[str] = Field(default=None, description="节点详细说明文本")
    children: list[TraceNode] = Field(default_factory=list, description="子节点列表")

    model_config = {"from_attributes": True}


class TraceResponse(BaseModel):
    """GET /api/explain/{module}/{record_id}/trace 响应"""
    module: str = Field(..., description="模块名：recognition / restoration")
    record_id: int = Field(..., description="记录 ID")
    root: TraceNode = Field(..., description="根节点：AI 输出结果的多维度分解树")
    disclaimer: str = Field(
        default="此图展示 AI 输出结果的多维度分解，非 AI 内部推理过程。"
                "Qwen-VL 等视觉模型为一次性调用返回所有信息，"
                "不存在「先判定品类→再判定年代→再判定产地」的逐步推理。",
        description="诚实声明"
    )


# ── 四色常量（前端同步） ──────────────────────────────────

SOURCE_COLORS: dict[SourceType, str] = {
    SourceType.COMPUTED: "#52c41a",      # 绿色
    SourceType.MODEL_OUTPUT: "#1890ff",  # 蓝色
    SourceType.RETRIEVED: "#fa8c16",     # 橙色
    SourceType.RULE_BASED: "#722ed1",    # 紫色
}

File: frontend/src/services/xai.ts
/** XAI 可解释体系 — API 客户端 + 类型定义 */

import api from './api'

// ── 来源类型（四色编码） ──

export type SourceType = 'computed' | 'model-output' | 'retrieved' | 'rule-based'

export const SOURCE_COLORS: Record<SourceType, string> = {
  computed: '#52c41a',       // 绿色 — 可验证计算
  'model-output': '#1890ff', // 蓝色 — AI 模型推断
  retrieved: '#fa8c16',      // 橙色 — 知识库检索
  'rule-based': '#722ed1',   // 紫色 — 规则判定
}

export const SOURCE_LABELS: Record<SourceType, string> = {
  computed: '计算判定',
  'model-output': '模型输出',
  retrieved: '知识检索',
  'rule-based': '规则判定',
}

// ── Trace 树节点 ──

export interface TraceNode {
  name: string
  source_type: SourceType
  confidence?: number | null
  detail?: string | null
  children: TraceNode[]
}

export interface TraceResponse {
  module: string
  record_id: number
  root: TraceNode
  disclaimer: string
}

// ── API 函数 ──

export async function fetchRecognitionTrace(recordId: number): Promise<TraceResponse> {
  const { data } = await api.get<TraceResponse>(`/api/explain/recognition/${recordId}/trace`)
  return data
}

export async function fetchRestorationTrace(recordId: number): Promise<TraceResponse> {
  const { data } = await api.get<TraceResponse>(`/api/explain/restoration/${recordId}/trace`)
  return data
}

// ── 工具函数 ──

/** 将 TraceResponse 转换为 ECharts tree 系列需要的格式 */
export function toEChartsTree(root: TraceNode) {
  function convert(node: TraceNode): any {
    const itemStyle = {
      color: SOURCE_COLORS[node.source_type] || '#666',
      borderColor: SOURCE_COLORS[node.source_type] || '#666',
    }
    const base: any = {
      name: node.name,
      itemStyle,
    }
    // 置信度附加到名称后
    if (node.confidence != null) {
      base.name = `${node.name} (${(node.confidence * 100).toFixed(0)}%)`
    }
    // tooltip 详情
    if (node.detail) {
      base.tooltip = {
        trigger: 'item',
        formatter: `{b}\n\n${node.detail}`,
      }
    }
    if (node.children && node.children.length > 0) {
      base.children = node.children.map(convert)
    } else {
      // ECharts tree 需要叶子节点有 value
      base.value = 1
    }
    return base
  }
  return convert(root)
}
